code/FsComposite/Composite.Core/Composite.fs

```diff
@@ -47,12 +45,22 @@ module CompositeToolset =
         | Cons(x, Nil) -> if x = obj then l x else [obj; x] |> LazyList.ofList
         | x -> LazyList.cons obj x
 
-//    let flat obj =
-//        let rec flatInner o =
-//            match o with
-//            | Composite x -> LazyList.collect flatInner x
-//            | Value x -> l x
-//
-//        match obj with
-//        | Value x -> obj
-//        | Composite x -> Composite (LazyList.map (fun y -> Composite(LazyList.map Value (flatInner y)) ) x)
\ No newline at end of file
+    let rec fold2 acc f_update_acc lst =
+        match acc with
+        |(None, _) | (_, None) -> match lst with
+                                    | Nil -> acc
+                                    | Cons(h, tail) -> fold2 (f_update_acc h acc) f_update_acc tail
+        | _ -> acc
+
+    let find_2_and_transform f_is_1 f_is_2 f_transform lst =
+        let f h t =
+            match f_is_1 h, f_is_2 h, t with
+            | _, true, (a, None) -> (a, Some h)
+            | true, _, (None, b) -> (Some h, b)
+            | _ -> t
+        match fold2 (None, None) f lst with
+        | x -> f_transform x
+
+    let find_2_and_transform_strict f_is_1 f_is_2 f_transform lst =
+        let f_transform_strict = function | (Some x1, Some x2) -> l(f_transform (x1, x2)) | _ ->  LazyList.empty
+        find_2_and_transform f_is_1 f_is_2 f_transform_strict lst   

```

Comment:
Arguably, the three functions above don't belong here. They deal with generic data which does not necessarily mean Composite instances. Moving to a separate file is advised.